### PR TITLE
Add alerts for DSP Operator

### PR DESCRIPTION
**Pull request by DharmitD**
## Description

Adding alerting rules for the Data Science Pipelines Operator.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] For commits that came from upstream, `[UPSTREAM]` has been prepended to the commit message.
- [x] JIRA link(s): https://issues.redhat.com/browse/RHODS-7876
- [x] The Jira story is acked.
- [ ] Live build image: 
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work.
- [ ] QE contact acknowledges that this has been tested and is approved for merge.

### CODE REVIEW

**Comment by DharmitD** [comment]
This PR is blocked until https://github.com/red-hat-data-services/odh-deployer/pull/314 and https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/merge_requests/82 are merged. 
Once they're merged, this PR will be rebased and marked ready for review.

**Comment by DharmitD** [comment]
> Remember to update the prometheus init-container (wait-for-deployment) in order to wait until the data science pipeline operator is active, or the alert will fire

Done, updated the prometheus init container.

**Comment by jgarciao** [comment]
Same for is needed for blackbox-exporter's init container (I just saw it)

File: monitoring/prometheus/blackbox-exporter-common.yaml [hunk 1]
```diff
@@ -72,6 +72,11 @@ spec:
               mm_svc_exists=false &&
               for i in `seq 1 230`; do sleep 10; echo "Waiting for modelmesh-controller service to become available..."; if curl -I http://modelmesh-controller.redhat-ods-applications.svc:8080/metrics; then mm_svc_exists=true; break; fi; done &&
               if [ $mm_svc_exists = false ]; then exit 1; fi; exit 0
+
+              dspo_svc_exists=false &&
+              for i in `seq 1 230`; do sleep 10; echo "Waiting for Data Science Pipelines Operator service to become available..."; if curl -sS -I http://data-science-pipelines-operator-service.redhat-ods-applications.svc:8080/metrics; then dspo_svc_exists=true; break; fi; done &&
+              if [ $dspo_svc_exists = false ]; then exit 1; fi; exit 0
```
Comment (harshad16) [inline monitoring/prometheus/blackbox-exporter-common.yaml:78]:
```suggestion
              if [ $mm_svc_exists = false ]; then exit 1; fi &&

              dspo_svc_exists=false &&
              for i in `seq 1 230`; do sleep 10; echo "Waiting for Data Science Pipelines Operator service to become available..."; if curl -sS -I http://data-science-pipelines-operator-service.redhat-ods-applications.svc:8080/metrics; then dspo_svc_exists=true; break; fi; done &&
              if [ $dspo_svc_exists = false ]; then exit 1; fi; exit 0
```

File: monitoring/prometheus/prometheus-configs.yaml [hunk 2]
```diff
@@ -600,6 +600,44 @@ data:
...
+             triage: "https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/blob/main/RHODS/Data-Science-Pipelines/data-science-pipelines-operator-probe-success-burn-rate.md"
+             summary: Data Science Pipelines Operator Probe Success Burn Rate
+           expr: |
+             sum(probe_success:burnrate2h{instance=~"data-science-pipelines-operator"}) by (instance) > (3.00 * (1-0.98000))
+             and
+             sum(probe_success:burnrate1d{instance=~"data-science-pipelines-operator"}) by (instance) > (3.00 * (1-0.98000))
+           for: 1h
+           labels:
+             severity: info
+>>>>>>> 3e9e7cc (Add alerts for DSPO)
```
Comment (harshad16) [inline monitoring/prometheus/prometheus-configs.yaml:639]:
```suggestion

```

File: monitoring/prometheus/prometheus.yaml [hunk 3]
```diff
@@ -48,6 +48,10 @@ spec:
               for i in `seq 1 230`; do sleep 10; echo "Waiting for modelmesh-controller service to become available..."; if curl -I http://modelmesh-controller.redhat-ods-applications.svc:8080/metrics; then mm_svc_exists=true; break; fi; done &&
               if [ $mm_svc_exists = false ]; then exit 1; fi; exit 0
 
+              dspo_svc_exists=false &&
+              for i in `seq 1 230`; do sleep 10; echo "Waiting for Data Science Pipelines Operator service to become available..."; if curl -sS -I http://data-science-pipelines-operator-service.redhat-ods-applications.svc:8080/metrics; then dspo_svc_exists=true; break; fi; done &&
+              if [ $dspo_svc_exists = false ]; then exit 1; fi; exit 0
```
Comment (harshad16) [inline monitoring/prometheus/prometheus.yaml:53]:
```suggestion
              if [ $mm_svc_exists = false ]; then exit 1; fi &&

              dspo_svc_exists=false &&
              for i in `seq 1 230`; do sleep 10; echo "Waiting for Data Science Pipelines Operator service to become available..."; if curl -sS -I http://data-science-pipelines-operator-service.redhat-ods-applications.svc:8080/metrics; then dspo_svc_exists=true; break; fi; done &&
              if [ $dspo_svc_exists = false ]; then exit 1; fi; exit 0
```

**Review by harshad16 — CHANGES_REQUESTED** [review]
Some changes are needed
Testing this further

File: monitoring/prometheus/prometheus.yaml (hunk 3 above)
Comment (DharmitD) [reply monitoring/prometheus/prometheus.yaml]: Done, updated to have this change.

File: monitoring/prometheus/prometheus-configs.yaml (hunk 2 above)
Comment (DharmitD) [reply monitoring/prometheus/prometheus-configs.yaml]: Updated to remove this.

File: monitoring/prometheus/blackbox-exporter-common.yaml (hunk 1 above)
Comment (DharmitD) [reply monitoring/prometheus/blackbox-exporter-common.yaml]: Done, updated to have this change.

**Comment by harshad16** [comment]
Tested the changes :
The alerts fires info level alerts
Screenshot from 2023-04-14 15-20-23
Screenshot from 2023-04-14 15-20-11

**Review by harshad16 — APPROVED** [review]
/lgtm

:+1:

File: monitoring/prometheus/prometheus-configs.yaml [hunk 4]
```diff
@@ -600,6 +600,43 @@ data:
...
+          for: 2m
+          labels:
+            severity: info
+        - alert: Data Science Pipelines Operator Probe Success Burn Rate
+          annotations:
+            message: 'High error budget burn for {{ $labels.instance }} (current value: {{ $value }}).'
+            triage: "https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/blob/main/RHODS/Data-Science-Pipelines/data-science-pipelines-operator-probe-success-burn-rate.md"
+            summary: Data Science Pipelines Operator Probe Success Burn Rate
+          expr: |
+            sum(probe_success:burnrate30m{instance=~"data-science-pipelines-operatorr"}) by (instance) > (6.00 * (1-0.98000))
```
Comment (harshad16) [inline monitoring/prometheus/prometheus-configs.yaml:621]:
```suggestion
            sum(probe_success:burnrate30m{instance=~"data-science-pipelines-operator"}) by (instance) > (6.00 * (1-0.98000))
```

**Review by harshad16 — CHANGES_REQUESTED** [review]
need one more change

File: monitoring/prometheus/prometheus-configs.yaml (hunk 4 above)
Comment (DharmitD) [reply monitoring/prometheus/prometheus-configs.yaml]: Done!

**Comment by harshad16** [comment]
Retested with changes:

Waited for all 3 alerts to fire:
Screenshot from 2023-04-14 15-48-10
Screenshot from 2023-04-14 15-47-44


Process for testing:

After applying the rules.
Scale down the service.
Alerts started firing.

/lgtm

**Comment by HumairAK** [comment]
From @jgarciao : 

> "If you are confident with PR319 and merge it, it will be included in a pre RC build DevOps will create on Monday (so I'll be able to test that with all the changes)" ~ Jorge


/label qe-approved

**Review by harshad16 — APPROVED** [review]
/lgtm

**Pull request merged**